src/types/Writable.ts

```diff
@@ -15,6 +15,18 @@ type WritableKeys<T> = {
 	[P in keyof T]: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, P>;
 }[keyof T];
 
+type OptionalKeys<T> = {
+	// map each attribute to either itself or never, depending on if it is optional
+	[P in keyof T]: {} extends Pick<T, P>
+		? P // If an empty Object type extends a type that only contains attribute P then we know attribute P is optional, since {} extends { a?: string } but not { a: string }
+		: never; // otherwise don't keep it
+}[keyof T]; // Pull out the mapped attributes (which are now exclusively either "never" or the attributes with optional types)
```

Comment:
These are great! Mind wrapping them?

Reply:
Will do -- though prettier ruins everything and puts half the comment after the `:` grumble grumble

Reply:
Ok fixed it using `/* */`

This is why I never use comments 😅